

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>lineage.resources &mdash; lineage 1.0b1.post.dev5 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../../genindex.html"/>
        <link rel="search" title="Search" href="../../search.html"/>
    <link rel="top" title="lineage 1.0b1.post.dev5 documentation" href="../../index.html"/>
        <link rel="up" title="lineage" href="../lineage.html"/> 

  
  <script src="../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../index.html" class="icon icon-home"> lineage
          

          
          </a>

          
            
            
              <div class="version">
                1.0b1.post.dev5
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../readme.html">README</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../output_files.html">Output Files</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">lineage</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
          <li><a href="../lineage.html">lineage</a> &raquo;</li>
        
      <li>lineage.resources</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for lineage.resources</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot; Class for downloading and loading required external resources.</span>

<span class="sd">`lineage` uses tables and data from UCSC&#39;s Genome Browser:</span>

<span class="sd">* http://genome.ucsc.edu/</span>
<span class="sd">* http://genome.ucsc.edu/cgi-bin/hgTables</span>

<span class="sd">References</span>
<span class="sd">----------</span>
<span class="sd">..[1] Karolchik D, Hinrichs AS, Furey TS, Roskin KM, Sugnet CW, Haussler D, Kent WJ.</span>
<span class="sd">  The UCSC Table Browser data retrieval tool. Nucleic Acids Res. 2004 Jan</span>
<span class="sd">  1;32(Database issue):D493-6. PubMed PMID: 14681465; PubMed Central PMCID:</span>
<span class="sd">  PMC308837. https://www.ncbi.nlm.nih.gov/pubmed/14681465</span>
<span class="sd">..[2] Tyner C, Barber GP, Casper J, Clawson H, Diekhans M, Eisenhart C, Fischer CM,</span>
<span class="sd">  Gibson D, Gonzalez JN, Guruvadoo L, Haeussler M, Heitner S, Hinrichs AS,</span>
<span class="sd">  Karolchik D, Lee BT, Lee CM, Nejad P, Raney BJ, Rosenbloom KR, Speir ML,</span>
<span class="sd">  Villarreal C, Vivian J, Zweig AS, Haussler D, Kuhn RM, Kent WJ. The UCSC Genome</span>
<span class="sd">  Browser database: 2017 update. Nucleic Acids Res. 2017 Jan 4;45(D1):D626-D634.</span>
<span class="sd">  doi: 10.1093/nar/gkw1134. Epub 2016 Nov 29. PubMed PMID: 27899642; PubMed Central</span>
<span class="sd">  PMCID: PMC5210591. https://www.ncbi.nlm.nih.gov/pubmed/27899642</span>
<span class="sd">..[3] International Human Genome Sequencing Consortium. Initial sequencing and</span>
<span class="sd">  analysis of the human genome. Nature. 2001 Feb 15;409(6822):860-921.</span>
<span class="sd">  http://dx.doi.org/10.1038/35057062</span>
<span class="sd">..[4] hg19 (GRCh37): Hiram Clawson, Brooke Rhead, Pauline Fujita, Ann Zweig, Katrina</span>
<span class="sd">  Learned, Donna Karolchik and Robert Kuhn, https://genome.ucsc.edu/cgi-bin/hgGateway?db=hg19</span>
<span class="sd">..[5] hg18 (NCBI36): Engineering effort led by Fan Hsu; QA effort led by Ann Zweig,</span>
<span class="sd">  https://genome.ucsc.edu/cgi-bin/hgGateway?db=hg18</span>

<span class="sd">&quot;&quot;&quot;</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Copyright (C) 2017 Andrew Riha</span>

<span class="sd">This program is free software: you can redistribute it and/or modify</span>
<span class="sd">it under the terms of the GNU General Public License as published by</span>
<span class="sd">the Free Software Foundation, either version 3 of the License, or</span>
<span class="sd">(at your option) any later version.</span>

<span class="sd">This program is distributed in the hope that it will be useful,</span>
<span class="sd">but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<span class="sd">MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the</span>
<span class="sd">GNU General Public License for more details.</span>

<span class="sd">You should have received a copy of the GNU General Public License</span>
<span class="sd">along with this program.  If not, see &lt;http://www.gnu.org/licenses/&gt;.</span>

<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">ftplib</span>
<span class="kn">import</span> <span class="nn">gzip</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">tarfile</span>
<span class="kn">import</span> <span class="nn">tempfile</span>
<span class="kn">import</span> <span class="nn">urllib.request</span>
<span class="kn">import</span> <span class="nn">zlib</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="kn">import</span> <span class="nn">lineage</span>


<div class="viewcode-block" id="Resources"><a class="viewcode-back" href="../../lineage.html#lineage.resources.Resources">[docs]</a><span class="k">class</span> <span class="nc">Resources</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Object used to manage resources required by `lineage`. &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">resources_dir</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Initialize a ``Resources`` object.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        resources_dir : str</span>
<span class="sd">            name / path of resources directory</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_resources_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">resources_dir</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_hapmap_h36</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_hapmap_h37</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_cytoband_h36</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_cytoband_h37</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_knownGene_h37</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_kgXref_h37</span> <span class="o">=</span> <span class="kc">None</span>

<div class="viewcode-block" id="Resources.get_hapmap_h36"><a class="viewcode-back" href="../../lineage.html#lineage.resources.Resources.get_hapmap_h36">[docs]</a>    <span class="k">def</span> <span class="nf">get_hapmap_h36</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Get International HapMap Consortium HapMap for Build 36.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        dict</span>
<span class="sd">            dict of pandas.DataFrame HapMap tables if loading was successful, else None</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_hapmap_h36</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_hapmap_h36</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_load_hapmap</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_get_path_hapmap_h36</span><span class="p">())</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_hapmap_h36</span></div>

<div class="viewcode-block" id="Resources.get_hapmap_h37"><a class="viewcode-back" href="../../lineage.html#lineage.resources.Resources.get_hapmap_h37">[docs]</a>    <span class="k">def</span> <span class="nf">get_hapmap_h37</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Get International HapMap Consortium HapMap for Build 37.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        dict</span>
<span class="sd">            dict of pandas.DataFrame HapMap tables if loading was successful, else None</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_hapmap_h37</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_hapmap_h37</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_load_hapmap</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_get_path_hapmap_h37</span><span class="p">())</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_hapmap_h37</span></div>

<div class="viewcode-block" id="Resources.get_cytoband_h36"><a class="viewcode-back" href="../../lineage.html#lineage.resources.Resources.get_cytoband_h36">[docs]</a>    <span class="k">def</span> <span class="nf">get_cytoband_h36</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Get UCSC cytoBand table for Build 36.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pandas.DataFrame</span>
<span class="sd">            cytoBand table if loading was successful, else None</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_cytoband_h36</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_cytoband_h36</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_load_cytoband</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_get_path_cytoband_h36</span><span class="p">())</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_cytoband_h36</span></div>

<div class="viewcode-block" id="Resources.get_cytoband_h37"><a class="viewcode-back" href="../../lineage.html#lineage.resources.Resources.get_cytoband_h37">[docs]</a>    <span class="k">def</span> <span class="nf">get_cytoband_h37</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Get UCSC cytoBand table for Build 37.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pandas.DataFrame</span>
<span class="sd">            cytoBand table if loading was successful, else None</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_cytoband_h37</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_cytoband_h37</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_load_cytoband</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_get_path_cytoband_h37</span><span class="p">())</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_cytoband_h37</span></div>

<div class="viewcode-block" id="Resources.get_knownGene_h37"><a class="viewcode-back" href="../../lineage.html#lineage.resources.Resources.get_knownGene_h37">[docs]</a>    <span class="k">def</span> <span class="nf">get_knownGene_h37</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Get UCSC knownGene table for Build 37.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pandas.DataFrame</span>
<span class="sd">            knownGene table if loading was successful, else None</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_knownGene_h37</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_knownGene_h37</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_load_knownGene</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_get_path_knownGene_h37</span><span class="p">())</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_knownGene_h37</span></div>

<div class="viewcode-block" id="Resources.get_kgXref_h37"><a class="viewcode-back" href="../../lineage.html#lineage.resources.Resources.get_kgXref_h37">[docs]</a>    <span class="k">def</span> <span class="nf">get_kgXref_h37</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Get UCSC kgXref table for Build 37.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pandas.DataFrame</span>
<span class="sd">            kgXref table if loading was successful, else None</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_kgXref_h37</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_kgXref_h37</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_load_kgXref</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_get_path_kgXref_h37</span><span class="p">())</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_kgXref_h37</span></div>

<div class="viewcode-block" id="Resources.download_example_datasets"><a class="viewcode-back" href="../../lineage.html#lineage.resources.Resources.download_example_datasets">[docs]</a>    <span class="k">def</span> <span class="nf">download_example_datasets</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Download example datasets from `openSNP &lt;https://opensnp.org&gt;`_.</span>

<span class="sd">        Per openSNP, &quot;the data is donated into the public domain using `CC0 1.0</span>
<span class="sd">        &lt;http://creativecommons.org/publicdomain/zero/1.0/&gt;`_.&quot;</span>

<span class="sd">        References</span>
<span class="sd">        ----------</span>
<span class="sd">        ..[1] Greshake B, Bayer PE, Rausch H, Reda J (2014), &quot;openSNP–A Crowdsourced Web Resource</span>
<span class="sd">          for Personal Genomics,&quot; PLOS ONE, 9(3): e89204,</span>
<span class="sd">          https://doi.org/10.1371/journal.pone.0089204</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_download_file</span><span class="p">(</span><span class="s1">&#39;https://opensnp.org/data/662.23andme.304&#39;</span><span class="p">,</span>
                            <span class="s1">&#39;662.23andme.304.csv.gz&#39;</span><span class="p">,</span> <span class="n">compress</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_download_file</span><span class="p">(</span><span class="s1">&#39;https://opensnp.org/data/662.23andme.340&#39;</span><span class="p">,</span>
                            <span class="s1">&#39;662.23andme.340.csv.gz&#39;</span><span class="p">,</span> <span class="n">compress</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_download_file</span><span class="p">(</span><span class="s1">&#39;https://opensnp.org/data/662.ftdna-illumina.341&#39;</span><span class="p">,</span>
                            <span class="s1">&#39;662.ftdna-illumina.341.csv.gz&#39;</span><span class="p">,</span> <span class="n">compress</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_download_file</span><span class="p">(</span><span class="s1">&#39;https://opensnp.org/data/663.23andme.305&#39;</span><span class="p">,</span>
                            <span class="s1">&#39;663.23andme.305.csv.gz&#39;</span><span class="p">,</span> <span class="n">compress</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># these two files consist of concatenated gzip files and therefore need special handling</span>
        <span class="n">gzip_paths</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">gzip_paths</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_download_file</span><span class="p">(</span><span class="s1">&#39;https://opensnp.org/data/4583.ftdna-illumina.3482&#39;</span><span class="p">,</span>
                                              <span class="s1">&#39;4583.ftdna-illumina.3482.csv.gz&#39;</span><span class="p">))</span>
        <span class="n">gzip_paths</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_download_file</span><span class="p">(</span><span class="s1">&#39;https://opensnp.org/data/4584.ftdna-illumina.3483&#39;</span><span class="p">,</span>
                                              <span class="s1">&#39;4584.ftdna-illumina.3483.csv.gz&#39;</span><span class="p">))</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">gzip_path</span> <span class="ow">in</span> <span class="n">gzip_paths</span><span class="p">:</span>
                <span class="c1"># https://stackoverflow.com/q/4928560</span>
                <span class="c1"># https://stackoverflow.com/a/37042747</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">gzip_path</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                    <span class="n">decompressor</span> <span class="o">=</span> <span class="n">zlib</span><span class="o">.</span><span class="n">decompressobj</span><span class="p">(</span><span class="mi">31</span><span class="p">)</span>

                    <span class="c1"># decompress data from first concatenated gzip file</span>
                    <span class="n">data</span> <span class="o">=</span> <span class="n">decompressor</span><span class="o">.</span><span class="n">decompress</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>

                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">decompressor</span><span class="o">.</span><span class="n">unused_data</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="c1"># decompress data from second concatenated gzip file, if any</span>
                        <span class="n">additional_data</span> <span class="o">=</span> <span class="n">zlib</span><span class="o">.</span><span class="n">decompress</span><span class="p">(</span><span class="n">decompressor</span><span class="o">.</span><span class="n">unused_data</span><span class="p">,</span> <span class="mi">31</span><span class="p">)</span>
                        <span class="n">data</span> <span class="o">+=</span> <span class="n">additional_data</span><span class="p">[</span><span class="mi">33</span><span class="p">:]</span>  <span class="c1"># skip over second header</span>

                <span class="c1"># recompress data</span>
                <span class="k">with</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">gzip_path</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">err</span><span class="p">)</span></div>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_load_hapmap</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Load HapMap.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filename : str</span>
<span class="sd">            path to compressed archive with HapMap data</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        hapmap : dict</span>
<span class="sd">            dict of pandas.DataFrame HapMap tables if loading was successful, else None</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">hapmap</span> <span class="o">=</span> <span class="p">{}</span>

            <span class="k">if</span> <span class="s1">&#39;36&#39;</span> <span class="ow">in</span> <span class="n">filename</span><span class="p">:</span>
                <span class="k">with</span> <span class="n">tarfile</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">tar</span><span class="p">:</span>
                    <span class="c1"># http://stackoverflow.com/a/2018576</span>
                    <span class="k">for</span> <span class="n">member</span> <span class="ow">in</span> <span class="n">tar</span><span class="o">.</span><span class="n">getmembers</span><span class="p">():</span>
                        <span class="k">if</span> <span class="s1">&#39;genetic_map&#39;</span> <span class="ow">in</span> <span class="n">member</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">tar</span><span class="o">.</span><span class="n">extractfile</span><span class="p">(</span><span class="n">member</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">)</span>
                            <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;position&#39;</span><span class="p">:</span> <span class="s1">&#39;pos&#39;</span><span class="p">,</span>
                                                    <span class="s1">&#39;COMBINED_rate(cM/Mb)&#39;</span><span class="p">:</span> <span class="s1">&#39;rate&#39;</span><span class="p">,</span>
                                                    <span class="s1">&#39;Genetic_Map(cM)&#39;</span><span class="p">:</span> <span class="s1">&#39;map&#39;</span><span class="p">})</span>
                            <span class="n">start_pos</span> <span class="o">=</span> <span class="n">member</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s1">&#39;chr&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">3</span>
                            <span class="n">end_pos</span> <span class="o">=</span> <span class="n">member</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s1">&#39;_b36&#39;</span><span class="p">)</span>
                            <span class="n">hapmap</span><span class="p">[</span><span class="n">member</span><span class="o">.</span><span class="n">name</span><span class="p">[</span><span class="n">start_pos</span><span class="p">:</span><span class="n">end_pos</span><span class="p">]]</span> <span class="o">=</span> <span class="n">df</span>
            <span class="k">elif</span> <span class="s1">&#39;37&#39;</span> <span class="ow">in</span> <span class="n">filename</span><span class="p">:</span>
                <span class="k">with</span> <span class="n">tarfile</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">tar</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">member</span> <span class="ow">in</span> <span class="n">tar</span><span class="o">.</span><span class="n">getmembers</span><span class="p">():</span>
                        <span class="k">if</span> <span class="s1">&#39;genetic_map&#39;</span> <span class="ow">in</span> <span class="n">member</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">tar</span><span class="o">.</span><span class="n">extractfile</span><span class="p">(</span><span class="n">member</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
                            <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;Position(bp)&#39;</span><span class="p">:</span> <span class="s1">&#39;pos&#39;</span><span class="p">,</span>
                                                    <span class="s1">&#39;Rate(cM/Mb)&#39;</span><span class="p">:</span> <span class="s1">&#39;rate&#39;</span><span class="p">,</span>
                                                    <span class="s1">&#39;Map(cM)&#39;</span><span class="p">:</span> <span class="s1">&#39;map&#39;</span><span class="p">})</span>
                            <span class="k">del</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">]</span>
                            <span class="n">start_pos</span> <span class="o">=</span> <span class="n">member</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s1">&#39;chr&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">3</span>
                            <span class="n">end_pos</span> <span class="o">=</span> <span class="n">member</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span>
                            <span class="n">hapmap</span><span class="p">[</span><span class="n">member</span><span class="o">.</span><span class="n">name</span><span class="p">[</span><span class="n">start_pos</span><span class="p">:</span><span class="n">end_pos</span><span class="p">]]</span> <span class="o">=</span> <span class="n">df</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">hapmap</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">err</span><span class="p">)</span>
            <span class="n">hapmap</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">return</span> <span class="n">hapmap</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_load_cytoband</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Load UCSC cytoBand table.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filename : str</span>
<span class="sd">            path to cytoBand file</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        df : pandas.DataFrame</span>
<span class="sd">            cytoBand table if loading was successful, else None</span>

<span class="sd">        References</span>
<span class="sd">        ----------</span>
<span class="sd">        ..[1] Ryan Dale, GitHub Gist,</span>
<span class="sd">          https://gist.github.com/daler/c98fc410282d7570efc3#file-ideograms-py</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># adapted from chromosome plotting code (see [1]_)</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="s1">&#39;gie_stain&#39;</span><span class="p">])</span>
            <span class="n">df</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">3</span><span class="p">:]</span>
            <span class="k">return</span> <span class="n">df</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">err</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_load_knownGene</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Load UCSC knownGene table.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filename : str</span>
<span class="sd">            path to knownGene file</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        df : pandas.DataFrame</span>
<span class="sd">            knownGene table if loading was successful, else None</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;strand&#39;</span><span class="p">,</span> <span class="s1">&#39;txStart&#39;</span><span class="p">,</span> <span class="s1">&#39;txEnd&#39;</span><span class="p">,</span>
                                                <span class="s1">&#39;cdsStart&#39;</span><span class="p">,</span> <span class="s1">&#39;cdsEnd&#39;</span><span class="p">,</span> <span class="s1">&#39;exonCount&#39;</span><span class="p">,</span> <span class="s1">&#39;exonStarts&#39;</span><span class="p">,</span>
                                                <span class="s1">&#39;exonEnds&#39;</span><span class="p">,</span> <span class="s1">&#39;proteinID&#39;</span><span class="p">,</span> <span class="s1">&#39;alignID&#39;</span><span class="p">],</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
            <span class="n">df</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">3</span><span class="p">:]</span>
            <span class="k">return</span> <span class="n">df</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">err</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_load_kgXref</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Load UCSC kgXref table.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filename : str</span>
<span class="sd">            path to kgXref file</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        df : pandas.DataFrame</span>
<span class="sd">            kgXref table if loading was successful, else None</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;kgID&#39;</span><span class="p">,</span> <span class="s1">&#39;mRNA&#39;</span><span class="p">,</span> <span class="s1">&#39;spID&#39;</span><span class="p">,</span> <span class="s1">&#39;spDisplayID&#39;</span><span class="p">,</span>
                                                <span class="s1">&#39;geneSymbol&#39;</span><span class="p">,</span> <span class="s1">&#39;refseq&#39;</span><span class="p">,</span> <span class="s1">&#39;protAcc&#39;</span><span class="p">,</span>
                                                <span class="s1">&#39;description&#39;</span><span class="p">,</span> <span class="s1">&#39;rfamAcc&#39;</span><span class="p">,</span> <span class="s1">&#39;tRnaName&#39;</span><span class="p">],</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                               <span class="n">dtype</span><span class="o">=</span><span class="nb">object</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">df</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">err</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">_get_path_cytoband_h36</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Get local path to cytoBand file for hg18 / NCBI36 from UCSC, downloading if necessary.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        str</span>
<span class="sd">            path to cytoband_h36.txt.gz</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_download_file</span><span class="p">(</span>
            <span class="s1">&#39;ftp://hgdownload.cse.ucsc.edu/goldenPath/hg18/database/cytoBand.txt.gz&#39;</span><span class="p">,</span>
            <span class="s1">&#39;cytoband_h36.txt.gz&#39;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_get_path_cytoband_h37</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Get local path to cytoBand file for hg19 / GRCh37 from UCSC, downloading if necessary.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        str</span>
<span class="sd">            path to cytoband_h37.txt.gz</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_download_file</span><span class="p">(</span>
            <span class="s1">&#39;ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/cytoBand.txt.gz&#39;</span><span class="p">,</span>
            <span class="s1">&#39;cytoband_h37.txt.gz&#39;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_get_path_hapmap_h36</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Get local path to HapMap for hg18 / NCBI36, downloading if necessary.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        str</span>
<span class="sd">            path to hapmap_h36.tar.gz</span>

<span class="sd">        References</span>
<span class="sd">        ----------</span>
<span class="sd">        ..[1] &quot;The International HapMap Consortium (2007).  A second generation human haplotype</span>
<span class="sd">          map of over 3.1 million SNPs.  Nature 449: 851-861.&quot;</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">lineage</span><span class="o">.</span><span class="n">create_dir</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_resources_dir</span><span class="p">):</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="n">hapmap</span> <span class="o">=</span> <span class="s1">&#39;hapmap_h36&#39;</span>
        <span class="n">destination</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_resources_dir</span><span class="p">,</span> <span class="n">hapmap</span> <span class="o">+</span> <span class="s1">&#39;.tar.gz&#39;</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">destination</span><span class="p">):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="c1"># make FTP connection to NCBI</span>
                <span class="k">with</span> <span class="n">ftplib</span><span class="o">.</span><span class="n">FTP</span><span class="p">(</span><span class="s1">&#39;ftp.ncbi.nlm.nih.gov&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ftp</span><span class="p">:</span>
                    <span class="n">ftp</span><span class="o">.</span><span class="n">login</span><span class="p">()</span>
                    <span class="n">ftp</span><span class="o">.</span><span class="n">cwd</span><span class="p">(</span><span class="s1">&#39;hapmap/recombination/2008-03_rel22_B36/rates&#39;</span><span class="p">)</span>

                    <span class="c1"># download each HapMap file and add to compressed tar</span>
                    <span class="k">with</span> <span class="n">tarfile</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">destination</span><span class="p">,</span> <span class="s1">&#39;w:gz&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_tar</span><span class="p">:</span>
                        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">ftp</span><span class="o">.</span><span class="n">nlst</span><span class="p">():</span>
                            <span class="k">if</span> <span class="s1">&#39;.txt&#39;</span> <span class="ow">in</span> <span class="n">filename</span><span class="p">:</span>
                                <span class="n">path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">destination</span><span class="p">,</span> <span class="n">hapmap</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">_print_download_msg</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>

                                <span class="c1"># open temp file, download HapMap file, close temp file</span>
                                <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">delete</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="k">as</span> <span class="n">fp</span><span class="p">:</span>
                                    <span class="n">ftp</span><span class="o">.</span><span class="n">retrbinary</span><span class="p">(</span><span class="s1">&#39;RETR &#39;</span> <span class="o">+</span> <span class="n">filename</span><span class="p">,</span> <span class="n">fp</span><span class="o">.</span><span class="n">write</span><span class="p">)</span>

                                <span class="c1"># add temp file to archive</span>
                                <span class="n">out_tar</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">fp</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">arcname</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">hapmap</span><span class="p">,</span> <span class="n">filename</span><span class="p">))</span>

                                <span class="c1"># remove temp file</span>
                                <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">fp</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                    <span class="n">ftp</span><span class="o">.</span><span class="n">quit</span><span class="p">()</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">err</span><span class="p">)</span>
                <span class="k">return</span> <span class="kc">None</span>

        <span class="k">return</span> <span class="n">destination</span>

    <span class="k">def</span> <span class="nf">_get_path_hapmap_h37</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Get local path to HapMap for hg19 / GRCh37, downloading if necessary.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        str</span>
<span class="sd">            path to hapmap_h37.tar.gz</span>

<span class="sd">        References</span>
<span class="sd">        ----------</span>
<span class="sd">        ..[1] &quot;The map was generated by lifting the HapMap Phase II genetic map from build 35 to</span>
<span class="sd">          GRCh37. The original map was generated using LDhat as described in the 2007 HapMap</span>
<span class="sd">          paper (Nature, 18th Sept 2007). The conversion from b35 to GRCh37 was achieved using</span>
<span class="sd">          the UCSC liftOver tool. Adam Auton, 08/12/2010&quot;</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_download_file</span><span class="p">(</span>
            <span class="s1">&#39;ftp://ftp.ncbi.nlm.nih.gov/hapmap/recombination/2011-01_phaseII_B37&#39;</span>
            <span class="s1">&#39;/genetic_map_HapMapII_GRCh37.tar.gz&#39;</span><span class="p">,</span> <span class="s1">&#39;hapmap_h37.tar.gz&#39;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_get_path_knownGene_h37</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Get local path to knownGene file for hg19 / GRCh37 from UCSC, downloading if necessary.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        str</span>
<span class="sd">            path to knownGene_h37.txt.gz</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_download_file</span><span class="p">(</span>
            <span class="s1">&#39;ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/knownGene.txt.gz&#39;</span><span class="p">,</span>
            <span class="s1">&#39;knownGene_h37.txt.gz&#39;</span><span class="p">)</span>


    <span class="k">def</span> <span class="nf">_get_path_kgXref_h37</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Get local path to kgXref file for hg19 / GRCh37 from UCSC, downloading if necessary.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        str</span>
<span class="sd">            path to kgXref_h37.txt.gz</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_download_file</span><span class="p">(</span>
            <span class="s1">&#39;ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/kgXref.txt.gz&#39;</span><span class="p">,</span>
            <span class="s1">&#39;kgXref_h37.txt.gz&#39;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_download_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">url</span><span class="p">,</span> <span class="n">filename</span><span class="p">,</span> <span class="n">compress</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Download a file to the resources folder.</span>

<span class="sd">        Download data from `url`, save as `filename`, and optionally compress with gzip.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        url : str</span>
<span class="sd">            URL to download data from</span>
<span class="sd">        filename : str</span>
<span class="sd">            name of file to save; if compress, ensure &#39;.gz&#39; is appended</span>
<span class="sd">        compress : bool</span>
<span class="sd">            compress with gzip</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        str</span>
<span class="sd">            path to downloaded file, None if error</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">lineage</span><span class="o">.</span><span class="n">create_dir</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_resources_dir</span><span class="p">):</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="k">if</span> <span class="n">compress</span> <span class="ow">and</span> <span class="n">filename</span><span class="p">[</span><span class="o">-</span><span class="mi">3</span><span class="p">:]</span> <span class="o">!=</span> <span class="s1">&#39;.gz&#39;</span><span class="p">:</span>
            <span class="n">filename</span> <span class="o">+=</span> <span class="s1">&#39;.gz&#39;</span>

        <span class="n">destination</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_resources_dir</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">destination</span><span class="p">):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_print_download_msg</span><span class="p">(</span><span class="n">destination</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">compress</span><span class="p">:</span>
                    <span class="n">open_func</span> <span class="o">=</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">open_func</span> <span class="o">=</span> <span class="nb">open</span>

                <span class="c1"># get file if it hasn&#39;t already been downloaded</span>
                <span class="c1"># http://stackoverflow.com/a/7244263</span>
                <span class="k">with</span> <span class="n">urllib</span><span class="o">.</span><span class="n">request</span><span class="o">.</span><span class="n">urlopen</span><span class="p">(</span><span class="n">url</span><span class="p">)</span> <span class="k">as</span> <span class="n">response</span><span class="p">,</span> <span class="n">open_func</span><span class="p">(</span><span class="n">destination</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                    <span class="n">data</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>  <span class="c1"># a `bytes` object</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">err</span><span class="p">)</span>
                <span class="k">return</span> <span class="kc">None</span>

        <span class="k">return</span> <span class="n">destination</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_print_download_msg</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Print download message.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        path : str</span>
<span class="sd">            path to file being downloaded</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Downloading &#39;</span> <span class="o">+</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">relpath</span><span class="p">(</span><span class="n">path</span><span class="p">))</span></div>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Andrew Riha.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../',
            VERSION:'1.0b1.post.dev5',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>